packages/test/src/test-integration-update.ts

```diff
@@ -0,0 +1,319 @@
+  // eslint-disable-next-line @typescript-eslint/no-floating-promises
+  wfHandle.executeUpdate(stateMutatingUpdate);
+  await new Promise((res) => setTimeout(res, 1000));
+
+  const worker = await createWorker();
+  await worker.runUntil(async () => {
+    // Worker receives activation: [doUpdate, startWorkflow]
+    const wfResult = await wfHandle.result();
+    t.deepEqual(wfResult, 'mutated-by-update');
+  });
+});
+
+test('Two Updates in first WFT', async (t) => {
+  const { createWorker, startWorkflow } = helpers(t);
+  const wfHandle = await startWorkflow(workflowWithUpdates);
+
+  // Race condition: wait long enough for the Updates to have been admitted, so
+  // that they are in the first WFT, along with startWorkflow.
+  // eslint-disable-next-line @typescript-eslint/no-floating-promises
+  wfHandle.executeUpdate(update, { args: ['1'] });
+  // eslint-disable-next-line @typescript-eslint/no-floating-promises
+  wfHandle.executeUpdate(doneUpdate);
+  await new Promise((res) => setTimeout(res, 1000));
+
+  const worker = await createWorker();
+  await worker.runUntil(async () => {
+    // Worker receives activation: [doUpdate, doUpdate, startWorkflow]. The
+    // updates initially lack a handler, are pushed to a buffer, and are
+    // executed when their handler is available.
+    const wfResult = await wfHandle.result();
+    t.deepEqual(wfResult, ['1', 'done', '$']);
+  });
+});
+
+// The following test would fail if the point at which the Update handler is
+// executed differed between first execution and replay (in that case, the
+// Update implementation would be violating workflow determinism).
+const earlyExecutedUpdate = wf.defineUpdate('earlyExecutedUpdate');
+const handlerHasBeenExecutedQuery = wf.defineQuery<boolean>('handlerHasBeenExecutedQuery');
+const openGateSignal = wf.defineSignal('openGateSignal');
+
+export async function updateReplayTestWorkflow(): Promise<boolean> {
+  let handlerHasBeenExecuted = false;
+  wf.setHandler(earlyExecutedUpdate, () => void (handlerHasBeenExecuted = true));
+  const handlerWasExecutedEarly = handlerHasBeenExecuted;
+
+  wf.setHandler(handlerHasBeenExecutedQuery, () => handlerHasBeenExecuted);
+
+  let gateOpen = false;
+  wf.setHandler(openGateSignal, () => void (gateOpen = true));
+  await wf.condition(() => gateOpen);
+
+  return handlerWasExecutedEarly;
+}
+
+test('Update handler is called at same point during first execution and replay', async (t) => {
+  const { createWorker, startWorkflow } = helpers(t);
+
+  // Start a Workflow and an Update of that Workflow.
+  const wfHandle = await startWorkflow(updateReplayTestWorkflow);
+  // Race condition: wait long enough for the Update to have been admitted, so
+  // that it is in the first WFT, along with startWorkflow.
+  // eslint-disable-next-line @typescript-eslint/no-floating-promises
+  wfHandle.executeUpdate(earlyExecutedUpdate);
+  await new Promise((res) => setTimeout(res, 1000));
```

Comment:
Race condition will be fixed by https://github.com/dandavison/temporalio-sdk-typescript/pull/1

Reply:
I've eliminated most of the deliberate race conditions in tests by using the `startDelay` option added by @antlai-temporal. See 8248aad90b69b3a5fbe159fcf3a2805f9d2cfe52. This is the same technique as used in the Go test suite, e.g. https://github.com/temporalio/sdk-go/blob/c43c47fddc24109de92748f67d2fa63573d3e2b2/test/integration_test.go#L2487